core/src/main/java/io/parsingdata/metal/data/Slice.java

```diff
@@ -26,37 +29,44 @@
 
     public final Source source;
     public final long offset;
-    private final byte[] data; // Private because array content is mutable.
-    public final int size;
+    public final BigInteger length;
 
-    public Slice(final Source source, final long offset, final byte[] data) {
+    public Slice(final Source source, final long offset, final BigInteger length) {
         this.source = checkNotNull(source, "source");
         this.offset = offset;
-        this.data = checkNotNull(data, "data");
-        this.size = data.length;
+        this.length = checkNotNull(length, "length");
     }
 
     public byte[] getData() {
-        return data.clone();
+        return getData(length);
+    }
+
+    public byte[] getData(final BigInteger limit) {
+        if (limit.compareTo(BigInteger.ZERO) < 0) { return new byte[] {}; }
```

Comment:
If the limit is 0 and thus equal to ZERO, then you can also return an empty byte array, right? Our is a limit of 0 used as unlimited? Then my question would be; do you want to get data of unlimited length?

Reply:
`0` means zero. The reason I don't want to check at that location and delegate it to the underlying `Source` implementation is because I want to throw an exception if the offset of the `Slice` is beyond the size of the `Source`.

In other words, suppose you have a `Source` of 100 bytes. If you make a `Slice` at offset=110 with whatever size, even reading 0 bytes from it will throw an exception.

It's a rather theoretical thing btw, because the `Slice` would normally never be created. It's more in case you're doing something manually with slices/sources and this is defensive programming.

Reply:
The reason I pass a limit of 0 on to the source is because that throws an exception on any size as long as the offset is outside the bounds. So I explicitly don't want to handle it here.

But your comment did make we wonder why we interpret a negative value to mean 0. And if it is to be handled that way, it should instead call the underlying source with 0 (to get the same output as calling it with 0, for example in the outside bounds case I describe above).

However, I think this would be wrong. So I've now substituted it with an exception: it's simply not allowed to call this method with a negative value. Furthermore, I did notice that what I described (an exception on any size outside bounds) is checked in all `Source` implementations, but it should probably be checked in the `Source`'s general `slice()` method. So to be continued!

Reply:
This will be further addressed in #210.